### PR TITLE
Added basic test coverage for nested tensor masking

### PR DESCRIPTION
**Pull request by izdeby**
Methods covered:
- gen_nested_tensor
- to_tensor_mask
- nested_tensor_from_tensor_mask

ToDo:
- cover edge cases
- cover error cases

### CODE REVIEW

File: test/test_nested_tensor_masking.py [hunk 1]
```diff
@@ -3,42 +3,309 @@
 import pdb
 import sys
 import torch
-import nestedtensor
+import nestedtensor as NT
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: If people do `import numpy as np` and `import torch as th` then why do `import nestedtensor as NT`?

I don't care though, doesn't make a difference.

File: test/test_nested_tensor_masking.py [hunk 2]
```diff
@@ -3,42 +3,309 @@
...
-        random.shuffle(values)
-        nt = nestedtensor.nested_tensor(values)
-        tensor, mask = nt.to_tensor_mask()
+        nt = utils.gen_nested_tensor(seed=1, nested_dim=2, tensor_dim=1, size_low=2, size_high=2)
+        self.assertTrue(NT.is_nested_tensor(nt))
+        self.assertEqual(nt.nested_dim(), 2)
+        self.assertEqual(nt.size(), (2, 2, 2))
+        self.assertEqual(nt[0].dim(), 2)
+        self.assertEqual(nt[0].size(), (2, 2))
+        self.assertEqual(nt[0][0].size(), torch.Size([2]))
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: Why not use the nestedtensor.NestedSize constructor?

Comment (cpuhrsch) [reply test/test_nested_tensor_masking.py]: and compare with .nested_size()

File: test/test_nested_tensor_masking.py [hunk 3]
```diff
@@ -3,42 +3,309 @@
...
 import unittest
 from utils import TestCase
 import random
 
 import utils
 
 class TestTensorMask(TestCase):
-    # TODO: test comparison operator fails if nested_size doesn't agree.
+    def test_gen_nested_tensor(self):
+        nt = utils.gen_nested_tensor(seed=1, nested_dim=1, tensor_dim=1, size_low=2, size_high=2)
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: For these kind of very small cases it would make sense to hardcode the values. It's more readable and there isn't much value for this kind of very basic and pedantic test.

File: test/test_nested_tensor_masking.py [hunk 4]
```diff
@@ -3,42 +3,309 @@
...
+        nt = utils.gen_nested_tensor(seed=1, nested_dim=2, tensor_dim=2, size_low=2, size_high=2)
+        self.assertTrue(NT.is_nested_tensor(nt))
+        self.assertEqual(nt.nested_dim(), 2)
+        self.assertEqual(nt.size(), (2, 2, 2, 2))
+        self.assertEqual(nt[0].dim(), 3)
+        self.assertEqual(nt[0].size(), (2, 2, 2))
+        self.assertEqual(nt[0][0].size(), torch.Size([2, 2]))
+        self.assertEqual(nt[0][0][0].size(), torch.Size([2]))
+
+        nt = utils.gen_nested_tensor(seed=1, nested_dim=2, tensor_dim=2, size_low=10, size_high=10)
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: Is this actually helpful? If it works with size_low=2

File: test/test_nested_tensor_masking.py (hunk 3 above)
Comment (cpuhrsch) [reply test/test_nested_tensor_masking.py]: Also, you can get more variation. gen_nested_tensor in essence has value for testing invariants such as f(nt(t)) == nt(f(t)) for certain types of functions f.

File: test/test_nested_tensor_masking.py [hunk 5]
```diff
@@ -3,42 +3,309 @@
...
+            ]),
+            NT.nested_tensor([
+                NT.nested_tensor([
+                    torch.tensor([3, 4]),
+                    torch.tensor([5, 6])
+                ])
+            ])
+        ])
+        
+        # mask_dim = None
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: I'd split up all these subcases into separate tests such as "test_to_tensor_mask_dim_1" or whatever is most descriptive. That'll make it easier to figure out what failed and what's covered. If you wanted to be even more descriptive you'd add some more comments on the purpose of each test.

File: test/test_nested_tensor_masking.py [hunk 6]
```diff
@@ -3,42 +3,309 @@
...
+        res_nt = NT.nested_tensor_from_tensor_mask(tensor, mask, nested_dim=2)
+        self.assertEqual(expected_nt2, res_nt)
+
+        res_nt = NT.nested_tensor_from_tensor_mask(tensor, mask, nested_dim=3)
+        self.assertEqual(original_nt, res_nt)
+
 
 if __name__ == "__main__":
-    unittest.main()
+    unittest.main()
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: Lint nit

File: test/test_nested_tensor_masking.py (hunk 3 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: what values would you hardcode here? we are testing if helper `gen_nested_tensor` works as expected.

File: test/test_nested_tensor_masking.py (hunk 4 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: removed.

File: test/test_nested_tensor_masking.py (hunk 2 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: Done.

File: test/test_nested_tensor_masking.py (hunk 1 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: good point. renamed.

File: test/test_nested_tensor_masking.py (hunk 5 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: Yeah, I was thinking about it. Agree, its cleaner. I think comments would be a bit overkill. Will add them where its not obvious what is being tested or what are the expectations.

File: test/test_nested_tensor_masking.py (hunk 3 above)
Comment (cpuhrsch) [reply test/test_nested_tensor_masking.py]: That sounds like it should be a separate test specific to gen_nested_tensor

File: test/test_nested_tensor_masking.py [hunk 7]
```diff
@@ -3,42 +3,288 @@
...
+        self.assertEqual(nt1.size(), (2, 2))
+        self.assertEqual(nt1[0].dim(), 1)
+        self.assertEqual(nt1[0].size(), torch.Size([2]))
+
+        nt1 = utils.gen_nested_tensor(seed=1, nested_dim=2, tensor_dim=1, size_low=2, size_high=2)
+        self.assertTrue(nt.is_nested_tensor(nt1))
+        self.assertEqual(nt1.nested_dim(), 2)
+        self.assertEqual(nt1.size(), (2, 2, 2))
+        self.assertEqual(nt1[0].dim(), 2)
+        self.assertEqual(nt1.nested_size(), nt.NestedSize([[[2],[2]], [[2],[2]]]))
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: After https://github.com/pytorch/nestedtensor/commit/02ef5f895a8c21f5f304d625c320b60c231df984 you need to use ```from utils import nested_size_to_list``` instead of the constructor or introduce the constructor.

File: test/test_nested_tensor_masking.py [hunk 8]
```diff
@@ -3,42 +3,288 @@
...
+                                               [False, False]]],
+                                             [[[ True, True],
+                                               [ True, True]]]]))
+        self.assertEqual(mask.size(), torch.Size([2, 1, 2, 2]))
+        self.assertEqual(mask.dim(), 4)
+        self.assertEqual(tensor.size(), torch.Size([2, 1, 2, 2]))
+        self.assertEqual(tensor.dim(), 4)
+        self.assertEqual(tensor.dim(), original_nt2.dim())
+
+    def test_to_tensor_mask_nt_dim_5(self):
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: with all this repetition it seems worth writing a local generator function or even something more abstract

Comment (izdeby) [reply test/test_nested_tensor_masking.py]: Ok, will refactor to be nicer. 
I personally prefer writing test code like this cause its immediately obvious what is failing and why. But i can see why this is not ideal.

Comment (izdeby) [reply test/test_nested_tensor_masking.py]: I still don't have a 100% trust for `assertEqual` and just wanted to make sure that all the expectations are correct and my understanding of the expected result is right. Will refactor but it would be nice if you could take a quick look at few random cases and verify the expectations.

File: test/test_nested_tensor_masking.py [hunk 9]
```diff
@@ -3,42 +3,257 @@
...
-        nt = nestedtensor.nested_tensor(values)
-        tensor, mask = nt.to_tensor_mask()
-        self.assertTrue([len(value) for value in values] == list(mask.sum(-1)))
+    def test_gen_nested_tensor(self):
+        nt1 = utils.gen_nested_tensor(seed=1, nested_dim=1, tensor_dim=1, size_low=2, size_high=2)
+        self.assertTrue(nt.is_nested_tensor(nt1))
+        self.assertEqual(nt1.nested_dim(), 1)
+        self.assertEqual(nt1.size(), (2, 2))
+        self.assertEqual(nt1[0].dim(), 1)
+        self.assertEqual(nt1[0].size(), torch.Size([2]))
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: you're not testing nested_size here

File: test/test_nested_tensor_masking.py [hunk 10]
```diff
@@ -3,42 +3,257 @@
...
+
+        nt1 = utils.gen_nested_tensor(seed=1, nested_dim=2, tensor_dim=1, size_low=2, size_high=2)
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: The good thing about parameterizing on the seed is that you'll always get the same numbers. so you can pick a higher range in size_low and size_high instead of just 2.

File: test/test_nested_tensor_masking.py [hunk 11]
```diff
@@ -3,42 +3,257 @@
...
+    #
+    # Group of tests to test to_tensor_mask() 
+    # 
+    def test_to_tensor_mask_nt_dim_2(self):
+        nt1 = nt.nested_tensor([
+            torch.tensor([1, 2]),
+            torch.tensor([3, 4])
+        ])
+        exp_tensor = torch.tensor([[1, 2], [3, 4]])
+        exp_mask = torch.tensor(True)
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: We might need to think about returning torch.tensor(true) vs. torch.tensor([true]). In what case would we expect torch.tensor(true) and not torch.tensor([true])? This is also then a necessary test case.

File: test/test_nested_tensor_masking.py [hunk 12]
```diff
@@ -3,42 +3,257 @@
...
 from utils import TestCase
 import random
-
 import utils
+from utils import nested_size_to_list
+
+#
+# Test data
+#
+original_nt = nt.nested_tensor([
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: I think there's two arguments to be made for inlining them

a) readability
b) allowing to explicitly test in-place vs. out-of-place changes. Can a mask every be a view? Should it be? etc.

File: test/test_nested_tensor_masking.py [hunk 13]
```diff
@@ -3,42 +3,257 @@
...
+        res_nt = nt.nested_tensor_from_tensor_mask(tensor, mask)
+        self.assertEqual(expected_nt0, res_nt)
+
+        res_nt = nt.nested_tensor_from_tensor_mask(tensor, mask, nested_dim=0)
+        self.assertEqual(expected_nt0, res_nt)
+
+        res_nt = nt.nested_tensor_from_tensor_mask(tensor, mask, nested_dim=1)
+        self.assertEqual(expected_nt1, res_nt)
+
+        res_nt = nt.nested_tensor_from_tensor_mask(tensor, mask, nested_dim=2)
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: Two nestedtensors with different nested_dim shouldn't be considered equal. Why does the result form this match that of all the other ones above?

Comment (cpuhrsch) [reply test/test_nested_tensor_masking.py]: It's probably correct to raise a ValueError if the passed mask doesn't allow the reconstruction of a NestedTensor with the specified nested_dim

File: test/test_nested_tensor_masking.py (hunk 12 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: Moved them into the tests

File: test/test_nested_tensor_masking.py (hunk 9 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: added a test

File: test/test_nested_tensor_masking.py (hunk 10 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: Updated seed for one of the test cases

File: test/test_nested_tensor_masking.py (hunk 11 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: added this as a weekly agenda item

File: test/test_nested_tensor_masking.py (hunk 13 above)
Comment (izdeby) [reply test/test_nested_tensor_masking.py]: Updated

**Comment by izdeby** [comment]
@cpuhrsch, should be good to go.

File: nestedtensor/nested/masking.py [hunk 14]
```diff
@@ -14,15 +14,16 @@ def nested_tensor_from_padded_tensor(tensor, nested_dim=None, padding=-1):
     mask = (tensor != padding)
     return nested_tensor_from_tensor_mask(tensor, mask, nested_dim)
 
+"""
```
Comment (cpuhrsch) [inline nestedtensor/nested/masking.py]: why did you move the docstring out?

File: nestedtensor/nested/masking.py [hunk 15]
```diff
@@ -14,15 +14,16 @@ def nested_tensor_from_padded_tensor(tensor, nested_dim=None, padding=-1):
...
-    """
-    Given a tensor t of size N_1 x N_2 x ... N_n and mask m of size M_1 x M_2 x ... M_d where d <= n
-    return a NestedTensor of dimensionality d where subtensor t[i_1][i_2]...[i_d] is included as a Tensor constituent if m[i_1][i_2]...[i_d].
-    The resulting NestedTensor will be of dimension n.
-    Unless specific, it is attempted to yield a NestedTensor of minimal nested dimension.
-    """
     # TODO: Should be possible with views only.
-    assert mask.dim() > 0
+    if not mask.dim() > 0:
+        raise RuntimeError("Mask has to have dimention > 0")
```
Comment (cpuhrsch) [inline nestedtensor/nested/masking.py]: I think technically this is supposed to be a ValueError too? Doesn't matter all that much.

File: nestedtensor/nested/masking.py [hunk 16]
```diff
@@ -44,6 +45,9 @@ def _mergable_tensors(tensors, nested_dim):
         else:
             return creation.nested_tensor(tensors)
 
+    if nested_dim is not None and mask.dim() < nested_dim:
+        raise ValueError("Mask dimension ({0}) is too small to construct nested tensor with nested dimension of {1}".format(mask.dim(), nested_dim))
```
Comment (cpuhrsch) [inline nestedtensor/nested/masking.py]: nit: NestedTensor is the name of the type

File: test/test_nested_tensor_masking.py [hunk 17]
```diff
@@ -3,42 +3,348 @@
...
+        exp_mask = torch.tensor([[[ True, False]],
+                                 [[ True, True]]])
+        exp_tensor = torch.tensor([[[[1, 2], [0, 0]]],
+                                   [[[3, 4], [5, 6]]]])
+
+        self.assertEqual(res_mask, exp_mask)
+        self.assertEqual(res_tensor, exp_tensor)
+
+        # mask_dim = 0
+        res_tensor, res_mask = original_nt2.to_tensor_mask(mask_dim=0)
```
Comment (cpuhrsch) [inline test/test_nested_tensor_masking.py]: There's a choice we can make here. We can raise an error since it's not constructible with that mask dimension or do what you did. I'd prefer an error (I know that changing the code you're testing isn't part of testing) because it increases identifiability.
